be/src/exec/except_node.cpp

```diff
@@ -56,4 +40,157 @@ Status ExceptNode::init(const TPlanNode& tnode, RuntimeState* state) {
+    RETURN_IF_ERROR(exec_debug_action(TExecNodePhase::OPEN));
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+    RETURN_IF_CANCELLED(state);
+    // open result expr lists.
+    for (const vector<ExprContext*>& exprs : _child_expr_lists) {
+        RETURN_IF_ERROR(Expr::open(exprs, state));
+    }
+    // initial build hash table
+    _hash_tbl.reset(new HashTable(_child_expr_lists[0], _child_expr_lists[1], _build_tuple_size,
+                                  true, _find_nulls, id(), mem_tracker(), 1024));
+    RowBatch build_batch(child(0)->row_desc(), state->batch_size(), mem_tracker());
+    RETURN_IF_ERROR(child(0)->open(state));
+
+    bool eos = false;
+    while (!eos) {
+        RETURN_IF_CANCELLED(state);
+        RETURN_IF_ERROR(child(0)->get_next(state, &build_batch, &eos));
+        // take ownership of tuple data of build_batch
+        _build_pool->acquire_data(build_batch.tuple_data_pool(), false);
+        RETURN_IF_LIMIT_EXCEEDED(state, " Except, while constructing the hash table.");
+        for (int i = 0; i < build_batch.num_rows(); ++i) {
+            _hash_tbl->insert(build_batch.get_row(i));
+        }
+        VLOG_ROW << "hash table content: " << _hash_tbl->debug_string(true, &child(0)->row_desc());
+        build_batch.reset();
+    }
+    // if a table is empty, the result must be empty
+
+    if (_hash_tbl->size() == 0) {
+        _hash_tbl_iterator = _hash_tbl->begin();
+        return Status::OK();
+    }
+
+    for (int i = 1; i < _children.size(); ++i) {
+        // probe
+        _probe_batch.reset(new RowBatch(child(i)->row_desc(), state->batch_size(), mem_tracker()));
+        RETURN_IF_ERROR(child(i)->open(state));
+        eos = false;
+        while (!eos) {
+            RETURN_IF_CANCELLED(state);
+            RETURN_IF_ERROR(child(i)->get_next(state, _probe_batch.get(), &eos));
+            RETURN_IF_LIMIT_EXCEEDED(state, " Except , while probing the hash table.");
+            for (int j = 0; j < _probe_batch->num_rows(); ++j) {
+                _hash_tbl_iterator = _hash_tbl->find(_probe_batch->get_row(j));
+                if (_hash_tbl_iterator != _hash_tbl->end()) {
+                    _hash_tbl_iterator.set_matched();
+                }
+            }
+            _probe_batch->reset();
+        }
+        if (i != _children.size() - 1) {
+            // if this hash table is probed rebuild it
+            std::unique_ptr<HashTable> temp_tbl(
+                    new HashTable(_child_expr_lists[0], _child_expr_lists[i], _build_tuple_size,
+                                  true, _find_nulls, id(), mem_tracker(), 1024));
+            _hash_tbl_iterator = _hash_tbl->begin();
+            uint32_t previous_hash = -1;
+            while (_hash_tbl_iterator.has_next()) {
+                if (previous_hash != _hash_tbl_iterator.get_hash()) {
+                    previous_hash = _hash_tbl_iterator.get_hash();
+                    if (!_hash_tbl_iterator.matched()) {
```

Comment:
why do you use  'if (!_hash_tbl_iterator.matched())'?
may be you should use ' if (_hash_tbl_iterator.matched())'
matched means exceptNode

Reply:
a excepte b , use a to build a hash table and use b to probe
 the result is  element in a and never mached